### PR TITLE
Add support for multiple beans in the Flyway and Liquibase endpoints

### PR DESCRIPTION
**Pull request by eddumelendez**
See gh-6610

### CODE REVIEW

File: spring-boot-actuator/src/main/java/org/springframework/boot/actuate/endpoint/FlywayEndpoint.java [hunk 1]
```diff
@@ -38,21 +38,25 @@
...
 		Assert.notNull(flyway, "Flyway must not be null");
 		this.flyway = flyway;
 	}
 
 	@Override
-	public List<FlywayMigration> invoke() {
-		List<FlywayMigration> migrations = new ArrayList<FlywayMigration>();
-		for (MigrationInfo info : this.flyway.info().all()) {
-			migrations.add(new FlywayMigration(info));
+	public List<List<FlywayMigration>> invoke() {
```
Comment (wilkinsona) [inline spring-boot-actuator/src/main/java/org/springframework/boot/actuate/endpoint/FlywayEndpoint.java]: Changing to `List<List<FlywayMigration>>` breaks backwards compatibility without any benefit. I think we should either keep to old format, or change to a Map<String, List<FlywayMigration>> where the string somehow indicates which DataSource the migrations were run against. The JDBC URL of the DB would be ideal, but I'm not sure we can get that.

File: spring-boot-actuator/src/main/java/org/springframework/boot/actuate/endpoint/LiquibaseEndpoint.java [hunk 2]
```diff
@@ -37,34 +38,39 @@
...
 	}
 
 	@Override
-	public List<Map<String, ?>> invoke() {
-		StandardChangeLogHistoryService service = new StandardChangeLogHistoryService();
-		try {
-			DatabaseFactory factory = DatabaseFactory.getInstance();
-			DataSource dataSource = this.liquibase.getDataSource();
-			JdbcConnection connection = new JdbcConnection(dataSource.getConnection());
+	public List<List<Map<String, ?>>> invoke() {
```
Comment (wilkinsona) [inline spring-boot-actuator/src/main/java/org/springframework/boot/actuate/endpoint/LiquibaseEndpoint.java]: I think my [other comment](https://github.com/spring-projects/spring-boot/pull/6613/files#r74556845) applies here too.

**Comment by eddumelendez** [comment]
Hi @wilkinsona if there any comments in this PR please let me know :)

**Comment by wilkinsona** [comment]
Thanks for the updates, @eddumelendez.

I wonder if we could enhance Boot's `DataSourcePoolMetadata` to provide access to the URL rather than having to establish a connection to the database. Please don't do anything about that just now as it might turn out to be a bad idea. I just wanted to make a note of it here so I didn't forget.

**Comment by snicoll** [comment]
> I wonder if we could enhance Boot's DataSourcePoolMetadata to provide access to the UR

@wilkinsona I had a look to that and I am not sure we could, really. We would need to  change the contract of `DataSourcePoolMetadataProvider` somehow. Sure we can abstract the url access in the implementation but that's something that would require to open a connection anyway.

File: spring-boot-actuator/src/main/java/org/springframework/boot/actuate/endpoint/FlywayEndpoint.java [hunk 3]
```diff
@@ -38,21 +42,32 @@
...
  */
 @ConfigurationProperties(prefix = "endpoints.flyway")
-public class FlywayEndpoint extends AbstractEndpoint<List<FlywayMigration>> {
+public class FlywayEndpoint extends AbstractEndpoint<Map<String, List<FlywayMigration>>> {
 
-	private final Flyway flyway;
+	private final List<Flyway> flyway;
 
-	public FlywayEndpoint(Flyway flyway) {
+	public FlywayEndpoint(List<Flyway> flyway) {
```
Comment (snicoll) [inline spring-boot-actuator/src/main/java/org/springframework/boot/actuate/endpoint/FlywayEndpoint.java]: You're breaking backward compat there. Can you please add the constructor with a single resource back and just call the other one with `Collections.singletonList` ?

File: spring-boot-actuator/src/main/java/org/springframework/boot/actuate/endpoint/FlywayEndpoint.java [hunk 4]
```diff
@@ -38,21 +42,32 @@
...
 	@Override
-	public List<FlywayMigration> invoke() {
-		List<FlywayMigration> migrations = new ArrayList<FlywayMigration>();
-		for (MigrationInfo info : this.flyway.info().all()) {
-			migrations.add(new FlywayMigration(info));
+	public Map<String, List<FlywayMigration>> invoke() {
+		Map<String, List<FlywayMigration>> migrations = new HashMap<String, List<FlywayMigration>>();
+		for (Flyway flyway : this.flyway) {
+			try {
+				DatabaseMetaData metaData = flyway.getDataSource().getConnection().getMetaData();
```
Comment (snicoll) [inline spring-boot-actuator/src/main/java/org/springframework/boot/actuate/endpoint/FlywayEndpoint.java]: You need to close the `Connection`.

File: spring-boot-actuator/src/main/java/org/springframework/boot/actuate/endpoint/LiquibaseEndpoint.java [hunk 5]
```diff
@@ -37,34 +41,45 @@
...
+					finally {
+						connection.close();
+					}
+				}
+				catch (DatabaseException ex) {
+					throw new IllegalStateException("Unable to get Liquibase changelog", ex);
+				}
 			}
-			finally {
-				connection.close();
```
Comment (snicoll) [inline spring-boot-actuator/src/main/java/org/springframework/boot/actuate/endpoint/LiquibaseEndpoint.java]: ditto...

File: spring-boot-actuator/src/main/java/org/springframework/boot/actuate/endpoint/LiquibaseEndpoint.java [hunk 6]
```diff
@@ -37,34 +41,45 @@
...
+		Map<String, List<Map<String, ?>>> services = new HashMap<String, List<Map<String, ?>>>();
+
+		for (SpringLiquibase liquibase : this.liquibase) {
+			StandardChangeLogHistoryService service = new StandardChangeLogHistoryService();
 			try {
-				Database database = factory.findCorrectDatabaseImplementation(connection);
-				return service.queryDatabaseChangeLogTable(database);
+				DatabaseMetaData metaData = liquibase.getDataSource().getConnection().getMetaData();
+				try {
+					DatabaseFactory factory = DatabaseFactory.getInstance();
```
Comment (snicoll) [inline spring-boot-actuator/src/main/java/org/springframework/boot/actuate/endpoint/LiquibaseEndpoint.java]: Not sure you need to initialize the factory every time there.

File: spring-boot-actuator/src/test/java/org/springframework/boot/actuate/autoconfigure/EndpointAutoConfigurationTests.java [hunk 7]
```diff
@@ -207,6 +212,20 @@ public void testFlywayEndpoint() {
 	}
 
 	@Test
+	public void testTwoFlywayEndpoint() {
```
Comment (snicoll) [inline spring-boot-actuator/src/test/java/org/springframework/boot/actuate/autoconfigure/EndpointAutoConfigurationTests.java]: typo, `Endpoints`

File: spring-boot-actuator/src/test/java/org/springframework/boot/actuate/autoconfigure/MvcEndpointPathConfigurationTests.java [hunk 8]
```diff
@@ -153,12 +155,12 @@ public ConditionEvaluationReport conditionEvaluationReport(
 
 		@Bean
 		public FlywayEndpoint flyway() {
-			return new FlywayEndpoint(new Flyway());
+			return new FlywayEndpoint(Arrays.asList(new Flyway()));
```
Comment (snicoll) [inline spring-boot-actuator/src/test/java/org/springframework/boot/actuate/autoconfigure/MvcEndpointPathConfigurationTests.java]: Not breaking backward compat means you don't need to change this no more.

**Comment by snicoll** [comment]
@eddumelendez did you have a chance to review my comments? If you don't have time, let us know and we'll rework the PR.

**Comment by eddumelendez** [comment]
Sure @snicoll, I will update the PR. Thanks for feedback

**Comment by eddumelendez** [comment]
@snicoll PR has been updated. Sorry, I just notice you reviewed this 25 days ago... looks like I didn't read that notification 😢

**Comment by snicoll** [comment]
Don't be, it could a different notification from Github or something. I've seen others not replying to that kind of requests.

**Comment by snicoll** [comment]
OK, that looks good except one thing I am not sure. The key is the URL of the database, so you end up with something in the json output

```json
{
  "jdbc:h2:mem:testdb": [
    {
      "type": "SQL",
      "checksum": 710039845,
      "version": "1",
      "description": "init",
      "script": "V1__init.sql",
      "state": "SUCCESS",
      "installedOn": 1482400309894,
      "executionTime": 6
    }
  ]
}
```

I see no other way to generate a sensible id for the datasource but perhaps the team knows.

**Comment by snicoll** [comment]
Actually @bclozel helped me to answer that question.

**Comment by snicoll** [comment]
Thank you @eddumelendez - Exposing the URL of the datasource is something we're not keen doing (the health endpoint doesn't do that). Besides, it's quite hard to "index" data based on the url.

For that  stuff we usually use the name of the bean. That's what I've done in 5d909a9

**Pull request closed without merging**